cdc/puller/ddl_puller.go

```diff
@@ -42,6 +49,154 @@ const (
+	tableInfo *model.TableInfo
+	columnID  int64
+	outputCh  chan *model.DDLJobEntry
+}
+
+func (p *ddlJobPullerImpl) Run(ctx context.Context) error {
+	g, ctx := errgroup.WithContext(ctx)
+
+	g.Go(func() error {
+		return p.puller.Run(ctx)
+	})
+
+	rawDDLCh := memory.SortOutput(ctx, p.puller.Output())
+	var ddlRawKV *model.RawKVEntry
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case ddlRawKV = <-rawDDLCh:
+		}
+		if ddlRawKV == nil {
+			continue
+		}
+		job, err := p.unmarshalDDL(ddlRawKV)
+		p.outputCh <- &model.DDLJobEntry{
+			Job:    job,
+			OpType: ddlRawKV.OpType,
+			CRTs:   ddlRawKV.CRTs,
+			Err:    err,
+		}
+	}
+}
+
+func (p *ddlJobPullerImpl) Output() <-chan *model.DDLJobEntry {
+	return p.outputCh
+}
+
+func (p *ddlJobPullerImpl) initJobTableMeta() error {
+	version, err := p.kvStorage.CurrentVersion(tidbkv.GlobalTxnScope)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	snap, err := kv.GetSnapshotMeta(p.kvStorage, version.Ver)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	dbInfos, err := snap.ListDatabases()
+	if err != nil {
+		return cerror.WrapError(cerror.ErrMetaListDatabases, err)
+	}
+
+	db, err := findDBByName(dbInfos, mysql.SystemDB)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	tbls, err := snap.ListTables(db.ID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	tableInfo, err := findTableByName(tbls, "tidb_ddl_job")
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	col, err := findColumnByName(tableInfo.Columns, "job_meta")
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	p.tableInfo = model.WrapTableInfo(db.ID, db.Name.L, 0, tableInfo)
+	p.columnID = col.ID
+	return nil
+}
+
+func (p *ddlJobPullerImpl) unmarshalDDL(rawKV *model.RawKVEntry) (*timodel.Job, error) {
+	if rawKV.OpType != model.OpTypePut {
+		return nil, nil
+	}
+	if p.tableInfo == nil && !entry.IsLegacyFormatJob(rawKV) {
+		err := p.initJobTableMeta()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	return entry.ParseDDLJob(p.tableInfo, rawKV, p.columnID)
+}
+
+func findDBByName(dbs []*timodel.DBInfo, name string) (*timodel.DBInfo, error) {
+	for _, db := range dbs {
+		if db.Name.L == name {
+			return db, nil
+		}
+	}
+
+	return nil, cerror.WrapError(cerror.ErrDDLSchemaNotFound, errors.Errorf("can't find schema %s", name))
+}
+
+func findTableByName(tbls []*timodel.TableInfo, name string) (*timodel.TableInfo, error) {
+	for _, t := range tbls {
+		if t.Name.L == name {
+			return t, nil
+		}
+	}
+
+	return nil, cerror.WrapError(cerror.ErrDDLSchemaNotFound, errors.Errorf("can't find table %s", name))
+}
+
+func findColumnByName(cols []*timodel.ColumnInfo, name string) (*timodel.ColumnInfo, error) {
+	for _, c := range cols {
+		if c.Name.L == name {
+			return c, nil
+		}
+	}
+
+	return nil, cerror.WrapError(cerror.ErrDDLSchemaNotFound, errors.Errorf("can't find column %s", name))
+}
+
+// NewDDLJobPuller create a new NewDDLJobPuller fetch event start from checkpointTs
+// and put into buf.
```

Comment:
```suggestion
// NewDDLJobPuller creates a new NewDDLJobPuller, which fetches events starting from checkpointTs
// and put into buf.
```
What do yo mean by "buf"?

Reply:
no buf, I just copy it from `ddlPullerImpl`, removed